pkg/distributor/receiver/shim.go

```diff
@@ -0,0 +1,270 @@
+		r.logger.Log("msg", "receiver start")
+		err := receiver.Start(ctx, r)
+		if err != nil {
+			return fmt.Errorf("error starting receiver %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Called after distributor is asked to stop via StopAsync.
+func (r *receiversShim) stopping(_ error) error {
+	// when shutdown is called on the receiver it immediately shuts down its connection
+	// which drops requests on the floor. at this point in the shutdown process
+	// the readiness handler is already down . so we are not receiving any more requests.
+	// sleep for 30 seconds to here to all pending requests to finish.
+	time.Sleep(30 * time.Second)
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancelFn()
+
+	errs := make([]error, 0)
+
+	for _, receiver := range r.receivers {
+		err := receiver.Shutdown(ctx)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return multierr.Combine(errs...)
+	}
+
+	return nil
+}
+
+// ConsumeLogs implements consumer.Logs
+func (r *receiversShim) ConsumeLogs(ctx context.Context, ld pdata.Logs) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "distributor.ConsumeLogs")
+	defer span.Finish()
+	req, err := parseLog(ld)
+	if err != nil {
+		r.logger.Log("msg", "pusher failed to parse log data", "err", err)
+	}
+
+	start := time.Now()
+	_, err = r.pusher.Push(ctx, req)
+	metricPushDuration.Observe(time.Since(start).Seconds())
+	if err != nil {
+		r.logger.Log("msg", "pusher failed to consume log data", "err", err)
+	}
+
+	return err
+}
+
+func parseLog(ld pdata.Logs) (*logproto.PushRequest, error) {
+	streams := make(map[string]*logproto.Stream)
+	rss := ld.ResourceLogs()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		ill := rs.InstrumentationLibraryLogs()
+		for i := 0; i < ill.Len(); i++ {
+			logs := ill.At(i).LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+				pLog := logs.At(k)
+				labels := parseLabel(rs.Resource().Attributes(), pLog.Attributes())
+				entry, err := parseEntry(pLog)
+				if err != nil {
+					return nil, err
+				}
+				if stream, ok := streams[labels]; ok {
+					stream.Entries = append(stream.Entries, *entry)
+					continue
+				}
+				// Add the entry as a new stream
+				streams[labels] = &logproto.Stream{
+					Labels:  labels,
+					Entries: []logproto.Entry{*entry},
+				}
+			}
+		}
+	}
+	req := &logproto.PushRequest{
+		Streams: make([]logproto.Stream, 0, len(streams)),
+	}
+
+	for _, stream := range streams {
+		req.Streams = append(req.Streams, *stream)
+	}
+
+	return req, nil
+}
+
+func parseEntry(pLog pdata.LogRecord) (*logproto.Entry, error) {
+	record := LokiLogRecord{
+		SeverityNumber: int32(pLog.SeverityNumber()),
+		SeverityText:   pLog.SeverityText(),
+		Name:           pLog.Name(),
+		Body:           pLog.Body().AsString(),
+		Flags:          pLog.Flags(),
+		TraceID:        pLog.TraceID().HexString(),
+		SpanID:         pLog.SpanID().HexString(),
+	}
+
+	data, err := json.Marshal(&record)
```

Comment:
Can we allow logfmt too ?

Reply:
👌，done.
![image](https://user-images.githubusercontent.com/9583245/153696051-897c5bd9-6a0e-4a83-a500-1c9d9e4addce.png)